source/extensions/http/stateful_session/cookie/cookie.h

```diff
@@ -52,20 +62,55 @@ class CookieBasedSessionStateFactory : public Envoy::Http::SessionStateFactory {
   }
 
 private:
-  absl::optional<std::string> parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+  std::pair<absl::optional<std::string>, bool>
+  parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+    bool use_old_style_encoding = false;
     const std::string cookie_value = Envoy::Http::Utility::parseCookieValue(headers, name_);
-    std::string address = Envoy::Base64::decode(cookie_value);
+    if (cookie_value.empty()) {
+      return std::make_pair(absl::nullopt, use_old_style_encoding);
+    }
+    const std::string decoded_value = Envoy::Base64::decode(cookie_value);
+    std::string address;
+
+    // Try to interpret the cookie as proto payload.
+    // Otherwise treat it as "old" style format, which is ip-address:port.
+    envoy::Cookie cookie;
+    if (cookie.ParseFromString(decoded_value)) {
+      address = cookie.address();
+      if (address.empty() || (cookie.expires() == 0)) {
+        return std::make_pair(absl::nullopt, use_old_style_encoding);
+      }
+
+      std::chrono::seconds expiry_time(cookie.expires());
+      auto now = std::chrono::duration_cast<std::chrono::seconds>(
+          (time_source_.monotonicTime()).time_since_epoch());
+      if (now > expiry_time) {
+        // Ignore the address extracted from the cookie. This will cause
+        // upstream cluster to select a new host and new cookie will be generated.
+        return std::make_pair(absl::nullopt, use_old_style_encoding);
+      }
+    } else {
+      // Treat this as "old" style cookie.
+      ENVOY_LOG_MISC(
+          warn, "Non-proto cookie format detected. This format will be rejected in the future.");
+      address = decoded_value;
+      use_old_style_encoding = true;
+    }
```

Comment:
A runtime flag should be added here to tell if old format cookie value can be accepted or not. We can ignore all the old format cookie value after the runtime flags is expired and removed.
Or we can just keep the old format support forever. Because the overhead is very low.

Reply:
If I add the runtime guard, backwards compatibility will be lost. As it was noted in https://github.com/envoyproxy/envoy/issues/26347#issuecomment-1503506078, "old" format should be supported without tweaking runtime flags.

Reply:
Keep the logic (old format cookie parsing) forever is ok to me.

Reply:
Yeah, but it will be used only by people which hacks the cookie. I mean will manually generate the cookie and will attach it to the initial request. You can achieve the same by attaching stateful_session header. Cookies should not be used in such way. Therefore I think we should deprecate it in the future.

Reply:
Ideally we should always generate new format cookie.

And keep the function of parsing old format cookie will ensure almost all users will not be effected (their existing cookies will not be broken) and this feature will be useful for test with minor addtional code.